ratpack-pac4j/src/main/java/ratpack/pac4j/internal/RatpackWebContext.java

```diff
@@ -92,23 +109,43 @@ public Object getSessionAttribute(String name) {
   }
 
   @Override
+  public Object getSessionIdentifier() {
+    return session.getSession().getId();
+  }
+
+  @Override
   public String getRequestMethod() {
     return request.getMethod().getName();
   }
 
   @Override
+  public String getRemoteAddr() {
+    return request.getRemoteAddress().getHostText();
+  }
+
+  @Override
   public void writeResponseContent(String responseContent) {
     this.responseContent = responseContent;
   }
 
   @Override
   public void setResponseStatus(int code) {
-    context.getResponse().status(code);
+    response.status(code);
   }
 
   @Override
   public void setResponseHeader(String name, String value) {
-    context.getResponse().getHeaders().set(name, value);
+    response.getHeaders().set(name, value);
+  }
+
+  @Override
+  public void setResponseCharacterEncoding(String encoding) {
+    response.contentType(encoding);
```

Comment:
Something like that, but I'll have to look at what the contract for this method is. It might have to be more intelligent in order to support setting the encoding first and then the type. If that's the case, we'll need to hold on to the encoding in this type and implicitly add it to the type when set.